README_eng.md

```diff
@@ -0,0 +1,299 @@
+
+## Enum TransactionOutput.data
+
+Each transaction output must carry a `data` field. This field is the designation of the purpose of the transaction, we must highlight the following at the moment:
+
+* Transfer Tokens or NFT
+* Token issueance
+* Burning token
+* NFT creation
+
+This field must be `Enum`. All of the `TransactionOutput` structure and related types ​​must be derived `Encode` and `Decode`. This will allow SCALE Codec to correctly decode data fields. That there is no confusion due to changes in data places, we must mark each element of the type with the index `#[codec(index = NUM)]`.
+
+And there is a requirement for testing, each version must be covered with tests that are guaranteed to break if someone changes the internal data in any version. 
+
+```rust
+    #[derive(Encode, Decode...)]
+    pub struct TransactionOutput<AccountId> {
+        #[codec(index = 1)]
+        pub(crate) value: Value,
+        #[codec(index = 2)]
+        pub(crate) destination: Destination<AccountId>,
+        #[codec(index = 3)]
+        pub(crate) data: Option<TxData>,
+    }
+
+    #[derive(Encode, Decode...)]
+    pub enum TxData {
+        // TransferToken data to another user. If it is a token, then the token data must also be transferred to the recipient. 
+        #[codec(index = 1)]
+        TransferTokenV1(..),
+        // A new token creation
+        #[codec(index = 2)]
+        TokenIssuanceV1(..),
+        // Burning a token or NFT
+        #[codec(index = 3)]
+        TokenBurnV1(..),
+        // A new NFT creation
+        #[codec(index = 4)]
+        NftMintV1(..),
+        ...
+        // We do not change the data in TransferTokenV1, we just create a new version 
+        #[codec(index = 5)]
+        TransferTokenV2(..),
+    }
+```
+
+## Detailed description of Enum TxData fields
+
+### TransferTokenV1
+---
+
+Transfer funds to another person in a given UTXO. To send MLT we will use `TokenID :: MLT`, which is actually zero. If the `token_id` is equal to the ID of the MLS-01 token, then the amount of token is transferred to the recipient. The commission is taken only in MLT. If the `token_id` is equal to the id of any NFT, then the data of this NFT should be transferred to the recipient without changing the creator field. The UTXO model itself allows you to determine the owner of the NFT. 
+
+```rust
+    pub enum TxData {
+        TransferTokenV1{
+            token_id: TokenID,
+            amount: Value,
+        }
+        ...
+    }
+```
+
+### TokenIssuanceV1
+---
+
+When issuing a new token, we specify the data for creating a new token in the transaction input, where the `token_id` is: 
+
+```rust
+    let token_id = hash(&(inputs[0].hash, inputs[0].index));
```

Comment:
One thing that probably needs specification is how we're gonna create a string out of this. We can either make it just hex, which is ugly and very long. We can use base58check to make it look good, which is OK but will make a long string. We can also crop the byte array then use base58check. Or we can do what bitcoin does. Use RIPEMD160 hash, then use base58check. Just something to think about.

Reply:
I prefer using `<T as frame_system::Config>::Hashing` that in fact BlakeTwo256, means we can change this line to:

```
 let token_id = BlakeTwo256::hash_of(&(&tx.inputs[0], &tx.inputs[0].index));
```

But I found out that we do have not an `index` field in inputs and it seems we have to add it too.

Reply:
Accepted, for the now we will use:
```
 let token_id = BlakeTwo256::hash_of(&tx.inputs[0]);
```

When we'll had a use case for the string representation we'll use the above approach.